### PR TITLE
DraftText for security Consideration Section

### PR DESCRIPTION
**Pull request by Wind4Greg**
Some initial text for security consideration sections based on ECDSA and key management standards documents.

----

The security (integrity/authenticity) of a verifiable credential signed by a digital signature algorithm is dependent on a number of factors including:

* the correct application of the signature algorithm to a verifiable credential (this specification)
* the choice of of signature algorithm (ECDSA) and its parameters (P-256, P-384)
* the correct implementation and usage of the signature algorithm particularly with respect to well known trouble spots
* the proper management of the private and public keys use for signing and verification

In the following sections we review these important points and point the readers to where to find additional information.

*More stuff in the PR*...

***
<a href="https://pr-preview.s3.amazonaws.com/Wind4Greg/vc-di-ecdsa/pull/11.html" title="Last updated on Jun 15, 2023, 6:06 PM UTC (ae82469)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-di-ecdsa/11/c4a8938...Wind4Greg:ae82469.html" title="Last updated on Jun 15, 2023, 6:06 PM UTC (ae82469)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -693,14 +706,93 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+        </p>
+      </section>
+      <section class="informative">
+        <h3>Key Management</h3>
+        <p>The security of the ECDSA algorithm is dependent on the quality and 
+protection of its <em>private signing key</em>. Guidance in the management of 
+cryptographic keys is a large subject and the reader is referred to 
+[[NIST-SP-800-57-Part-1]] for more extensive recommendations and discussion. 
+As mentioned in both [[FIPS-186-5]] and [[NIST-SP-800-57-Part-1]] an ECDSA 
+private signing key should <strong>not</strong> be used for any other purpose
```
Comment (msporny) [inline index.html:778]: Avoid RFC "should", "must", and "may" language, even if lower cased, as reviewers get confused and tend to ask if we really meant SHOULD. Try alternative english language like "is expected to"... "it is strongly advised that"... "might" and so on.

File: index.html [hunk 2]
```diff
@@ -693,14 +706,93 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+        <p>The security of the ECDSA algorithm is dependent on the quality and 
+protection of its <em>private signing key</em>. Guidance in the management of 
+cryptographic keys is a large subject and the reader is referred to 
+[[NIST-SP-800-57-Part-1]] for more extensive recommendations and discussion. 
+As mentioned in both [[FIPS-186-5]] and [[NIST-SP-800-57-Part-1]] an ECDSA 
+private signing key should <strong>not</strong> be used for any other purpose 
+than ECDSA signatures.
+        </p>
+        <p>ECDSA private signing keys and <em>public verification keys</em> 
+should have limited <em>cryptoperiods</em> [[NIST-SP-800-57-Part-1]]. Where a
```
Comment (msporny) [inline index.html:782]: Same comment as above... use a different word than "should" here, please.

**Review by msporny — APPROVED** [review]
Two very minor changes requested, other than that, LGTM. Thank you for this, super helpful!

File: index.html [hunk 3]
```diff
@@ -693,14 +706,93 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
       </section>
     </section>
-    <section>
+    <section class="informative">
       <h2>Security Considerations</h2>
-      <p>
-The following section describes security considerations that developers
-implementing this specification should be aware of in order to create secure
-software.
+      <p>The security (integrity/authenticity) of a verifiable credential
```
Comment (davidlehn) [inline index.html:711]: The style of the rest of the doc is to have `<p>` and `<li>` start tags on their own line.

File: index.html (hunk 1 above)
Comment (Wind4Greg) [reply index.html]: Fixed. Thanks for wording recommendations.

File: index.html (hunk 2 above)
Comment (Wind4Greg) [reply index.html]: Fixed.

File: index.html (hunk 3 above)
Comment (Wind4Greg) [reply index.html]: Fixed.

File: index.html [hunk 4]
```diff
@@ -693,14 +706,105 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+a key size of 384 bits. [[NIST-SP-800-186]] recommendations includes curves 
+P-256 and P-384 at these security strength levels respectively.
+        </p>
+      </section>
+      <section class="informative">
+        <h3>Implementation Considerations for ECDSA Algorithms</h3>
+        <p>
+The ECDSA algorithm as detailed in [[FIPS-186-5]] states: &quot;A 
+new secret random number <em>k</em>, shall be generated prior to the generation 
+of each digital signature for use during the signature generation process.&quot;
```
Comment (TallTed) [inline index.html:770]: If you're going to quote, you must QUOTE. In the source document, the sentence reads:

> A new secret random number <em>k</em>, 0 < <em>k</em> < <em>n</em>, <strong>shall</strong> be generated prior to the generation of each digital signature for use during the signature generation process.

Based on the above *actual* quotation, our document should have either this, _with_ the range expression (note there is a comma both following `<em>k</em>` and preceding `<strong>shall</strong>`) —
```suggestion
The ECDSA algorithm as detailed in [[FIPS-186-5]] states: &quot;A 
new secret random number <em>k</em>, 0 < <em>k</em> < <em>n</em>,
<strong>shall</strong> be generated prior to the generation 
of each digital signature for use during the signature generation process.&quot; 
```

Comment (TallTed) [reply index.html:770]: — or this, _without_ the range expression but with the `...` indicating an omission from within the quoted material (note there is no comma following `<em>k</em>` nor preceding `<strong>shall</strong>`) —
```suggestion
The ECDSA algorithm as detailed in [[FIPS-186-5]] states: &quot;A 
new secret random number <em>k</em> ...
<strong>shall</strong> be generated prior to the generation 
of each digital signature for use during the signature generation process.&quot; 
```

File: index.html [hunk 5]
```diff
@@ -693,14 +706,105 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+      </section>
+      <section class="informative">
+        <h3>Implementation Considerations for ECDSA Algorithms</h3>
+        <p>
+The ECDSA algorithm as detailed in [[FIPS-186-5]] states: &quot;A 
+new secret random number <em>k</em>, shall be generated prior to the generation 
+of each digital signature for use during the signature generation process.&quot; 
+The failure to properly generate this <em>k</em> value has lead to some highly 
+public, integrity breaches in widely deployed systems. To counter this problem 
+a hash based method of determining the secret number <em>k</em>, called
```
Comment (TallTed) [inline index.html:773]:
```suggestion
The failure to properly generate this <em>k</em> value has lead to some highly 
publicized integrity breaches in widely deployed systems. To counter this problem, 
a hash-based method of determining the secret number <em>k</em>, called 
```

File: index.html [hunk 6]
```diff
@@ -693,14 +706,105 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+        </p>
+      </section>
+      <section class="informative">
+        <h3>Key Management</h3>
+        <p>
+The security of the ECDSA algorithm is dependent on the quality and 
+protection of its <em>private signing key</em>. Guidance in the management of 
+cryptographic keys is a large subject and the reader is referred to 
+[[NIST-SP-800-57-Part-1]] for more extensive recommendations and discussion. 
+As strongly recommended in both [[FIPS-186-5]] and [[NIST-SP-800-57-Part-1]] an ECDSA
```
Comment (TallTed) [inline index.html:787]:
```suggestion
As strongly recommended in both [[FIPS-186-5]] and [[NIST-SP-800-57-Part-1]], an ECDSA 
```

File: index.html [hunk 7]
```diff
@@ -693,14 +706,105 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+As strongly recommended in both [[FIPS-186-5]] and [[NIST-SP-800-57-Part-1]] an ECDSA 
+private signing key is not to be used for any other purpose 
+than ECDSA signatures.
+        </p>
+        <p>
+ECDSA private signing keys and <em>public verification keys</em> are strongly
+advised to have limited <em>cryptoperiods</em> [[NIST-SP-800-57-Part-1]]. Where 
+a <em>cryptoperiod</em> is &quot;the time span during which a specific key is 
+authorized for use by legitimate entities or the keys for a given system will 
+remain in effect.&quot;&quot; [[NIST-SP-800-57-Part-1]] gives extensive
```
Comment (TallTed) [inline index.html:796]:
```suggestion
advised to have limited <em>cryptoperiods</em> [[NIST-SP-800-57-Part-1]], where
a <em>cryptoperiod</em> is &quot;the time span during which a specific key is
authorized for use by legitimate entities or the keys for a given system will
remain in effect.&quot; [[NIST-SP-800-57-Part-1]] gives extensive
```

File: index.html [hunk 8]
```diff
@@ -693,14 +706,105 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+remain in effect.&quot;&quot; [[NIST-SP-800-57-Part-1]] gives extensive 
+guidance on cryptoperiods for different key types under different situations 
+and generally recommends a 1-3 year cryptoperiod for a private signing key.
+        </p>
+        <p>
+To deal with potential private key compromise [[NIST-SP-800-57-Part-1]] 
+gives recommendations for protective measure, harm reduction, and revocation. 
+Although we have be emphasizing the security of the private signing key, 
+Assurance of public-key validity is higly recommended on all public keys 
+before using them per [[NIST-SP-800-57-Part-1]].
```
Comment (TallTed) [inline index.html:805]:
```suggestion
To deal with potential private key compromises, [[NIST-SP-800-57-Part-1]] 
gives recommendations for protective measures, harm reduction, and revocation. 
Although we have been emphasizing the security of the private signing key, 
assurance of public key validity is highly recommended on all public keys 
before using them, per [[NIST-SP-800-57-Part-1]].
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Corrections and clarity tweaks

**Comment by msporny** [comment]
Merge conflicts, please fix.

**Comment by Wind4Greg** [comment]
?

File: index.html [hunk 9]
```diff
@@ -731,8 +731,8 @@ <h2>Security Considerations</h2>
         </li>
       </ul>
       <p>
-In the following sections we review these important points and point 
-the readers to where to find additional information.
+In the following sections, we review these important points and point
```
Comment (TallTed) [inline index.html:734]:
```suggestion
In the following sections, we review these important points and direct
```

File: index.html [hunk 10]
```diff
@@ -717,14 +730,106 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+including:
       </p>
-
+      <ul>
+        <li>
+the correct application of the signature algorithm to a verifiable
+credential (this specification)
+        </li>
+        <li>
+the choice of of signature algorithm (ECDSA) and its parameters (P-256, P-384)
```
Comment (dlongley) [inline index.html:746]:
```suggestion
the choice of signature algorithm (ECDSA) and its parameters (P-256, P-384)
```

File: index.html [hunk 11]
```diff
@@ -717,14 +730,106 @@ <h4>Proof Verification (ecdsa-2019)</h4>
...
+        </li>
+        <li>
+the correct implementation and usage of the signature algorithm,
+particularly with respect to well known problem areas
```
Comment (dlongley) [inline index.html:750]:
```suggestion
particularly with respect to well-known problem areas
```

**Review by dlongley — APPROVED** [review]

**Comment by msporny** [comment]
Editorial, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**